### PR TITLE
Marketplace Contract

### PR DESCRIPTION
**Pull request by calvinmd**
### Implemented
- Added deposit via onERC721Received callback
- Added withdrawal
- Added buy with Eth and buy with PLAT

### TODOs
- Hook up fees
- Integration test
- Test scripts

### CODE REVIEW

File: contracts/BitGuildFeeOracle.sol [hunk 1]
```diff
@@ -0,0 +1,53 @@
+pragma solidity ^0.4.24;
+
+import "./BitGuildAccessAdmin.sol";
+
+
+/**
+ * @title BitGuildFeeOracle
+ * @dev Fee definition
+ */
+contract BitGuildFeeOracle is BitGuildAccessAdmin {
```
Comment (vernon99) [inline contracts/BitGuildFeeOracle.sol]: See all my comments on this file from the previous PR. Not account in this one.

File: contracts/BitGuildMarketplace.sol [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+
+import "./BitGuildToken.sol";
+import "./BitGuildAccessAdmin.sol";
+import "./BitGuildWhitelist.sol";
+import "./BitGuildFeeOracle.sol";
+import "./SafeMath.sol";
+
+
+// @dev Interface for ERC-721 NFT
+contract ERC721Interface {
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: Move out in a separate file, import file. We will replace it later with a proper ERC721 impl when we will merge the repos.

File: contracts/BitGuildMarketplace.sol [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+interface ERC721TokenReceiver {
+	function onERC721Received(address _from, uint256 _tokenId, bytes data) external returns(bytes4);
+	function onERC721Received(address _operator, address _from, uint256 _tokenId, bytes data) external returns(bytes4);
+}
+
+/*
+ * @title BitGuildMarketplace
+ * @dev: Marketplace smart contract for BitGuild.com
+ */
+contract BitGuildMarketplace is BitGuildAccessAdmin, BitGuildWhitelist {
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: You definitely do not want to inherit it from the whitelist! You want to have a changeable whitelist link!

File: contracts/BitGuildMarketplace.sol [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+    function () external payable {
+        revert();
+    }
+
+    // @dev return all available item ids
+    function getAllItemIds() public view returns(uint[]) {
+        // @dev Push is not supported for memory. storage costs more gas. Use manual counter instead
+        uint[] memory tokens;
+        uint counter = 0;
+        for (uint i = 1; i <= currentItemId; i++) {
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: This is crazy and will die on any significant number of items. You cannot have a loop over all items.

File: contracts/BitGuildMarketplace.sol [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+        _;
+    }
+
+    // @dev fall back function
+    function () external payable {
+        revert();
+    }
+
+    // @dev return all available item ids
+    function getAllItemIds() public view returns(uint[]) {
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: See below, crazy, why do you even need it?

File: contracts/BitGuildMarketplace.sol [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    bytes4 constant ERC721_RECEIVED = 0x150b7a02;
+
+    // Predefined PLAT token
+    // BitGuildToken public token = BitGuildToken(0x7E43581b19ab509BCF9397a2eFd1ab10233f27dE); // Main Net
+    BitGuildToken public PLAT = BitGuildToken(0x0F2698b7605fE937933538387b3d6Fec9211477d); // Rinkeby
+
+    uint public defaultExpiry = 7 days;   // default expiry 7 days
+    // TODO: add function to modify this length 
+
+    uint public currentItemId = 1; // auto-increment
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: This is not good

File: contracts/BitGuildMarketplace.sol [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+    mapping(uint256 => address) public itemIdToContract;       // game contract the item belongs to
+    mapping(uint256 => address) public itemIdToSeller;         // seller of the item. 0x0 if sold/delisted
+    mapping(uint256 => uint256) public itemIdToTokenId;        // tokenId for this item
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: Why do you need the item id altogether?

File: contracts/BitGuildMarketplace.sol [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+    mapping(uint256 => uint256) public itemIdToPLAT;           // price in PLAT for this item
+    mapping(uint256 => uint256) public itemIdToETH;            // price in ETH for this item
+    mapping(uint256 => uint256) public itemIdToCreatedAt;      // timestamp when item is listed
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: All of that basically needs to be replaced with itemHash to orderData mapping. With the latter being a structure with price, currency enum, createdAt.

File: contracts/BitGuildFeeProvider.sol [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        uint oldPercentFee = customFee[key];
+        customFee[key] = _newFee;
+
+        emit LogCustomFeeChanged(_newFee, oldPercentFee, _buyer, _seller, _token, msg.sender);
+    }
+
+    /**
+     * @dev Calculate the custom fee based on buyer, seller, game token or combo of these
+     */
+    function getFee(uint _price, address _buyer, address _seller, address _token) public view returns(uint percent, uint fee) {
```
Comment (vernon99) [inline contracts/BitGuildFeeProvider.sol]: You need to also pass the currency! I suggest to do it as a token address (which can be PLAT address, 0 for eth, some other token address potentially). The param can be ignored for now, but we need to make sure we support it in the marketplace call.

File: contracts/BitGuildMarketplace.sol [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+import "./BitGuildAccessAdmin.sol";
+import "./BitGuildWhitelist.sol";
+import "./BitGuildFeeProvider.sol";
+import "./SafeMath.sol";
+import "./ERC721.sol";
+
+
+// @title ERC-721 Non-Fungible Token Standard
+// @dev Include interface for both new and old functions
+interface ERC721TokenReceiver {
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: Nit: Move to ERC721Receiver file or right to ERC721

File: contracts/BitGuildMarketplace.sol [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+    // BitGuildWhitelist public Whitelist = BitGuildWhitelist(); // Main Net
+    // BitGuildFeeProvide public FeeProvider = BitGuildFeeProvider(); // Main Net
+    BitGuildToken public PLAT = BitGuildToken(0x0F2698b7605fE937933538387b3d6Fec9211477d); // Rinkeby
+    BitGuildWhitelist public Whitelist = BitGuildWhitelist(0x72b93A4943eF4f658648e27D64e9e3B8cDF520a6); // Rinkeby
+    BitGuildFeeProvider public FeeProvider = BitGuildFeeProvider(0x9bE88C776299795A4996D351215274F3f1d84100); // Rinkeby
+
+    uint public defaultExpiry = 7 days;             // default expiry is 7 days
+    bool public allowWithdrawBeforeExpiry = false;  // allow withdraw listing before expiry? default is false
+
+    enum Currencies { PLAT, ETH }
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: Nit: probably singular, not plural (a standard convention, afaik). Ie Currency.

File: contracts/BitGuildFeeProvider.sol [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        bytes32 key = _getHash(_buyer, _seller, _token);
+        uint customPercentFee = customFee[key];
+        (percent, fee) = _getFee(_price, customPercentFee);
+    }
+
+    /**
+     * @dev Return the default fee
+     * @param _price for calculating fee
+     */
+    function getFee(uint _price) public view returns(uint percent, uint fee) {
```
Comment (vernon99) [inline contracts/BitGuildFeeProvider.sol]: I don't think you want this one to be exposed at all. The whole idea behind the extended version is for marketplace to pass needed parameters. What's the benefit of having a limited version?

File: contracts/BitGuildMarketplace.sol [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+    // @dev Retrieve hashkey to view listing
+    function getHashKey(address _contract, uint _tokenId) public pure returns(bytes32 key) {
+        key = _getHashKey(_contract, _tokenId);
+    }
+
+    // ===========================================
+    // Fee functions (from fee provider contract)
+    // ===========================================
+    // @dev get default fee
+    function getFee(uint _price) public view returns(uint percent, uint fee) {
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: Same here, I don't think you want the default like at all - that could lead to misuse

File: contracts/BitGuildMarketplace.sol [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+    // @dev Withdraw item from marketplace back to seller
+    // @param _contract whitelisted contract
+    // @param _tokenId  tokenId
+    function withdrawItem(address _contract, uint _tokenId) public onlyWhitelisted(_contract) {
+        bytes32 key = _getHashKey(_contract, _tokenId);
+        address seller = listings[key].seller;
+
+        require(seller == msg.sender, "Only seller can withdraw listing.");
+
+        if (!allowWithdrawBeforeExpiry) {
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: I think safe to remove the param and the logic here, we don't wanna lock the item before it expires

File: contracts/BitGuildMarketplace.sol [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+        emit LogItemSold(msg.sender, seller, _contract, _tokenId, price, currency, now);
+    }
+
+    // Buy with PLAT requires calling BitGuildToken contract, this is the callback
+    // call to approve already verified the token ownership, no checks required
+    // @param _buyer     buyer
+    // @param _value     PLAT amount (big number)
+    // @param _PLAT      BitGuild token address
+    // @param _extraData address _gameContract, uint _tokenId
+    function receiveApproval(address _buyer, uint _value, BitGuildToken _PLAT, bytes _extraData) public {
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: This is VERY uncool. A lot of dupe code! Abstract away. The difference here is like in 2-3 lines of code

File: contracts/BitGuildMarketplace.sol [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+        PLAT.transfer(msg.sender, balance);
+    }
+
+    // ===========================================
+    // Internal Functions
+    // ===========================================
+    // @dev clear listing for hash key
+    function _delist(bytes32 _key) internal {
+        // remove listing
+        listings[_key].currency = Currencies(0);
```
Comment (vernon99) [inline contracts/BitGuildMarketplace.sol]: I think there should be a single method to delete a mapping entry. This is fragile, imagine you add a new param and forget to update here.

File: contracts/BitGuildMarketplace.sol (hunk 15 above)
Comment (calvinmd) [reply contracts/BitGuildMarketplace.sol]: Abstracted out.

**Pull request merged**